lib/cn/cn_tree.c

```diff
@@ -1443,22 +1483,168 @@ cn_comp_update_spill(struct cn_compaction_work *work, struct kvset **kvsets)
 }
 
 /**
- * cn_comp_commit_spill() - commit spill operation to cndb log
- * See section comment for more info.
+ * cn_comp_update_split() - update tree after a node split operation
  */
 static merr_t
-cn_comp_commit_spill(struct cn_compaction_work *work, struct kvset **kvsets)
+cn_comp_update_split(
+    struct cn_compaction_work *w,
+    struct kvset *const       *kvsets,
+    const uint64_t            *nodeidv)
 {
-    struct cn_tree *         tree = work->cw_tree;
+    struct cn_tree *tree = w->cw_tree;
+    struct cn_tree_node *pnode = w->cw_node;
+    struct kvset_list_entry *le, *tmp;
+    struct list_head retired_kvsets;
+    struct cn_tree_node *left = NULL, *right = NULL;
+    char rekey[HSE_KVS_KEY_LEN_MAX];
+    uint k = 0, reklen;
+    merr_t err = 0;
 
-    /* Precondition: n_outputs == tree fanout */
-    assert(work->cw_outc == tree->ct_fanout);
-    if (ev(work->cw_outc != tree->ct_fanout))
-        return merr(EBUG);
+    if (ev(w->cw_err))
+        return w->cw_err;
 
-    /* Update tree and stats.  No failure paths allowed after ACK_C. */
-    cn_comp_update_spill(work, kvsets);
-    return 0;
+    INIT_LIST_HEAD(&retired_kvsets);
+
+    /* Allocate a new left node and add the split output kvsets on the left to this node.
+     * This need not be done under the tree lock as this new node is not published yet.
+     */
+    if (nodeidv[0] != CN_TREE_INVALID_NODEID) {
+        left = cn_node_alloc(tree, nodeidv[0]);
+        if (!left)
+            return merr(ENOMEM);
+
+        for (k = 0; k < w->cw_kvset_cnt; k++) {
+            if (kvsets[k])
+                kvset_list_add_tail(kvsets[k], &left->tn_kvset_list);
+        }
+
+        w->cw_split.nodev[0] = left;
+    }
+
+    if (nodeidv[1] != CN_TREE_INVALID_NODEID) {
+        /* pnode is protected by an exclusive compaction token, so the max key cannot
+         * change while a node split is in progress
+         */
+        cn_tree_node_get_max_key(pnode, rekey, sizeof(rekey), &reklen);
+    }
+
+    rmlock_wlock(&tree->ct_lock);
+    {
+        /* Move all kvsets from parent node to the retired list.
+         */
+        for (k = 0; k < w->cw_kvset_cnt; k++) {
+            assert(!list_empty(&pnode->tn_kvset_list));
+            le = list_last_entry(&pnode->tn_kvset_list, struct kvset_list_entry, le_link);
+            list_del(&le->le_link);
+            list_add(&le->le_link, &retired_kvsets);
+        }
+        assert(list_empty(&pnode->tn_kvset_list));
```

Comment:
Can you just replace this entire loop with list_splice(&pnoe->tn_kvset_list, &retired_kvsets) ?

Reply:
Good idea, totally forgot about splice. Thanks!